src/librustc_mir/borrow_check/nll/type_check/mod.rs

```diff
@@ -771,6 +771,22 @@ impl<'a, 'gcx, 'tcx> TypeChecker<'a, 'gcx, 'tcx> {
                     );
                 };
             }
+            StatementKind::UserAssertTy(ref ty, ref local) => {
+                let local_ty = mir.local_decls()[*local].ty;
+                debug!("check_stmt: user_assert_ty ty={:?} local_ty={:?}", ty, local_ty);
+                if let Err(terr) =
+                    self.eq_types(ty, local_ty, location.at_successor_within_block())
```

Comment:
I think this should be `at_self_within_block`, though it probably doesn't matter. But this basically means that the type must be equal "on entry" to the statement, which feels "intuitively" right to me.